### DISCUSSION TITLE
[Android] - Strange things happen with a custom configuration and data connection only

### DISCUSSION THREAD

**Discussion by d3cim** [DNSCrypt/dnscrypt-proxy#2020] [Potential issues]
## *What* is affected by this bug?
My Android rooted devices (no matter the Android version)

## *When* does this occur?
- The device boot up with dnscrypt-proxy installed on it.
- `routes` are configured with 2 relays at least.
- `skip_incompatible = true`.
- Wi-fi connection disabled.
- Data connection only.

## *How* do we replicate the issue?
On a Android rooted device with [Magisk](https://github.com/topjohnwu/Magisk) installed:
1. Disable wi-fi connection.
2. Enable data connection.
3. Download my [latest module release](https://github.com/quindecim/dnscrypt-proxy-android/releases/download/2.1.1-1/dnscrypt-proxy-android-v2.1.1-1.zip).
4. `Magisk > Modules > Install from storage > dnscrypt-proxy-android-v2.1.1-1.zip`.
5. Reboot.
6. Unlock the device (PIN and/or lock screen).
7. See the issue happen: it take almost 10 minutes to initialize the connection and make the X sign under LTE icon disappear. (sometimes after 10 minutes the connection isn't initialized yet).

## Expected behavior (i.e. solution)
The device should connect immediatly, just like under wi-fi only network.

1. I found useful (but again in random ways) toggling on/off the airplane mode.
2. I found useful toggling on the wi-fi connection. At this point the dnscrypt-proxy initialize the connection almost immediatly and then I can use data connection as well if I switch back to it.
3. `skip_incompatible = false` (not considered due to the lose of the routes correct functionality)

## Other Comments
This is the [dnscrypt-proxy.toml](https://github.com/quindecim/dnscrypt-proxy-android/blob/master/config/dnscrypt-proxy.toml) used.

Take logs was very difficult due to the fact that the issue happen randomly.
At the actual state, with no-wifi connection, is an high risk use dnscrypt-proxy with this configuration.

`log_level = 0`

[dnscrypt-proxy-wifi-only.log](https://github.com/DNSCrypt/dnscrypt-proxy/files/8009993/dnscrypt-proxy-wifi-only.log)

[dnscrypt-proxy-data-only.log](https://github.com/DNSCrypt/dnscrypt-proxy/files/8010014/dnscrypt-proxy-data-only.log)
[dnscrypt-proxy-data-only.jpg](https://user-images.githubusercontent.com/49964366/152678435-c5265276-12ba-47cb-965e-f86d87b6ebb2.jpg)

**Comment by Gedsh** [comment]
Have you tried force_tcp = true ?

**Reply by d3cim** [reply]
I did about 7 reboots without the issue using force_tcp = true, the initial connection is really fast, I will keep the configuration and do further tests.

As far as the issue happens randomly, and happen that dnscrypt-proxy can manage the initial connection with UDP queries as well, do you think is there anything that can be done to fix the issue without forcing TCP connections?

**Comment by Gedsh** [comment]
> do you think is there anything that can be done to fix the issue without forcing TCP connections?

I don't know Golang but there is a func _dnsExchange https://github.com/DNSCrypt/dnscrypt-proxy/blob/ef1c70e87da17576cf5ddef66957173884f5b48d/dnscrypt-proxy/dnsutils.go#L383

It checks if the network supports the fragmented queries required for the relay to work. Only @jedisct1 can know if this algorithm is not working correctly or if LTE networks simply do not support fragmented queries over udp.
